### PR TITLE
Fix: udder & wooly reagent creation

### PR DESCRIPTION
**Pull request by Sirionaut**
## About the PR
Fixes a minor bug I introduced in December: animals would keep consuming nutrients for reagent production even if the solution container was full.

Should also help with animal breeding since they can stay above the needed hungerthreshold without constant feeding.

## Why / Balance
A bit more feedback when feeding animals should make animal husbandry less opaque and more clear what to do for players

## Technical details
- Moved wooly an udder stuff to shared- Added a check for the `SolutionContainers` in udder and wooly system to be not full when consuming nutrients for reagent creation.
- `UdderSystem` and `WoolySystem` use `SharedSolutionContainerSystem` now
- `EgglayerSystem` uses `_timing.CurTime` instead of `AccumulatedFrametime`
- Added a bit of examine text for udder having entities to show players a rough hunger level (overfed, hungry, no hunger).
- egglaying has a slightly randomized `NextGrowth` timer

## Media

cowexamine
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- fix: no more nutrient consumption when udder/woolcoat is full
- add: udder having entities display rough hunger level when examined

### CODE REVIEW

File: Content.Server/Animals/Components/EggLayerComponent.cs [hunk 1]
```diff
@@ -9,44 +11,47 @@ namespace Content.Server.Animals.Components;
...
-[RegisterComponent]
+[RegisterComponent, Access(typeof(EggLayerSystem))]
 public sealed partial class EggLayerComponent : Component
 {
-    [DataField]
-    public EntProtoId EggLayAction = "ActionAnimalLayEgg";
+    /// <summary>
+    ///     The item that gets laid/spawned, retrieved from animal prototype.
+    /// </summary>
+    [DataField(required: true), ViewVariables(VVAccess.ReadWrite)]
```
Comment (Tayrtahn) [inline Content.Server/Animals/Components/EggLayerComponent.cs:20]:
```suggestion
    [DataField(required: true)]
```
DataField automatically gives VVAccess.ReadWrite

This goes for a bunch of other DataFields in this PR too

File: Content.Server/Animals/Components/EggLayerComponent.cs [hunk 2]
```diff
@@ -9,44 +11,47 @@ namespace Content.Server.Animals.Components;
...
+    [DataField, ViewVariables(VVAccess.ReadWrite)]
+    public float HungerUsage = 25f;
 
     [DataField] public EntityUid? Action;
+
+    /// <summary>
+    ///     When to next try to produce.
+    /// </summary>
+    [DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), ViewVariables(VVAccess.ReadWrite)]
+    public TimeSpan NextGrowth = TimeSpan.FromSeconds(0);
```
Comment (Tayrtahn) [inline Content.Server/Animals/Components/EggLayerComponent.cs:56]: This should use AutoPausedField with AutoGenerateComponentPause on the component.

File: Content.Server/Animals/Systems/EggLayerSystem.cs [hunk 3]
```diff
@@ -60,11 +66,12 @@ public override void Update(float frameTime)
     private void OnMapInit(EntityUid uid, EggLayerComponent component, MapInitEvent args)
     {
         _actions.AddAction(uid, ref component.Action, component.EggLayAction);
-        component.CurrentEggLayCooldown = _random.NextFloat(component.EggLayCooldownMin, component.EggLayCooldownMax);
+        component.NextGrowth = TimeSpan.FromSeconds(_random.NextFloat(component.EggLayCooldownMin, component.EggLayCooldownMax));
```
Comment (Tayrtahn) [inline Content.Server/Animals/Systems/EggLayerSystem.cs:69]: MapInit runs when the entity spawns, so this should use curtime + the random float, otherwise, after a few minutes have passed in the round, they're going to lay eggs immediately on spawning.

File: Content.Server/Animals/Systems/UdderSystem.cs [hunk 4]
```diff
@@ -1,8 +1,9 @@
 using Content.Server.Animals.Components;
```
Comment (Tayrtahn) [inline Content.Server/Animals/Systems/UdderSystem.cs:1]: This seems like it would be pretty easy to move to Shared and add prediction.

File: Content.Server/Animals/Systems/WoolySystem.cs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 using Content.Server.Animals.Components;
```
Comment (Tayrtahn) [inline Content.Server/Animals/Systems/WoolySystem.cs:1]: This also seems easy enough to move to Shared

File: Resources/Locale/en-US/animals/udder/udder-system.ftl [hunk 6]
```diff
@@ -5,3 +5,7 @@ udder-system-success = You fill {THE($target)} with {$amount}u from the udder.
 udder-system-dry = The udder is dry.
 
 udder-system-verb-milk = Milk
+
+udder-system-examine-overfed = This animal looks overfed
+udder-system-examine-hungry = This animal looks hungry
+udder-system-examine-none = This animal seems to not get hungry
```
Comment (Tayrtahn) [inline Resources/Locale/en-US/animals/udder/udder-system.ftl:11]: Split infinitive! 

```suggestion
udder-system-examine-none = This animal seems not to get hungry
```

Or


```suggestion
udder-system-examine-none = This animal seems like it doesn't get hungry
```

Comment (Tayrtahn) [inline Resources/Locale/en-US/animals/udder/udder-system.ftl:11]: These might be nicer as `{ SUBJECT($entity) } looks overfed` (etc), passing in the entity when doing the Loc.GetString call.

File: Content.Server/Animals/Systems/UdderSystem.cs [hunk 7]
```diff
@@ -134,4 +139,20 @@ private void AddMilkVerb(Entity<UdderComponent> entity, ref GetVerbsEvent<Altern
...
+        {
+            if (_hunger.GetHungerThreshold(hunger) == HungerThreshold.Overfed)
+                args.PushMarkup(Loc.GetString("udder-system-examine-overfed"));
+
+            if (_hunger.GetHungerThreshold(hunger) <= HungerThreshold.Peckish)
+                args.PushMarkup(Loc.GetString("udder-system-examine-hungry"));
+        }
+        else
+        {
+            args.PushMarkup(Loc.GetString("udder-system-examine-none"));
```
Comment (Tayrtahn) [inline Content.Server/Animals/Systems/UdderSystem.cs:155]: This might be more elegant as a [relational switch expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#relational-patterns), but it's fine like this too. Just a comment.

**Review by Tayrtahn — CHANGES_REQUESTED** [review]

File: Content.Server/Animals/Components/EggLayerComponent.cs (hunk 2 above)
Comment (Sirionaut) [reply Content.Server/Animals/Components/EggLayerComponent.cs:56]: only EggLayerComponent or the others as well?

Comment (Tayrtahn) [reply Content.Server/Animals/Components/EggLayerComponent.cs:56]: The others too, yeah.

**Comment by Sirionaut** [comment]
So it works, mostly, can't figure out why it now deletes goats after fully eating their wool.

**Comment by Tayrtahn** [comment]
I cleaned up some of the move to Shared. If the Client and Server versions are going to be identical, you don't need to have them anymore, you can just make the Shared version sealed.
That fixed the goat-eating problem too.

File: Content.Server/Animals/Components/EggLayerComponent.cs [hunk 8]
```diff
@@ -9,44 +10,47 @@ namespace Content.Server.Animals.Components;
...
+[RegisterComponent, Access(typeof(EggLayerSystem)), AutoGenerateComponentPause]
 public sealed partial class EggLayerComponent : Component
 {
-    [DataField]
-    public EntProtoId EggLayAction = "ActionAnimalLayEgg";
+    /// <summary>
+    ///     The item that gets laid/spawned, retrieved from animal prototype.
+    /// </summary>
+    [DataField(required: true)]
+    public List<EntitySpawnEntry> EggSpawn = default!;
```
Comment (metalgearsloth) [inline Content.Server/Animals/Components/EggLayerComponent.cs:20]:
```suggestion
    [DataField(required: true)]
    public List<EntitySpawnEntry> EggSpawn = new();
```

This is just going to throw needlessly if someone uses addcomp.

File: Content.Server/Animals/Components/EggLayerComponent.cs [hunk 9]
```diff
@@ -9,44 +10,47 @@ namespace Content.Server.Animals.Components;
...
+    [DataField]
+    public float EggLayCooldownMin = 45f;
 
     /// <summary>
     ///     Maximum cooldown used for the automatic egg laying.
     /// </summary>
-    [DataField, ViewVariables(VVAccess.ReadWrite)]
-    public float EggLayCooldownMax = 120f;
+    [DataField]
+    public float EggLayCooldownMax = 90f;
```
Comment (metalgearsloth) [inline Content.Server/Animals/Components/EggLayerComponent.cs:41]: I'm not a fan of making large balance changes inside of bugfix PRs for 2 reasons:
1. If we need to revert the change it makes it much nastier.
2. It's much more difficult for forks to pull 1 without the other as they need to manually fix it themselves.

File: Content.Server/Animals/Systems/EggLayerSystem.cs [hunk 10]
```diff
@@ -39,19 +40,24 @@ public override void Update(float frameTime)
...
 
-            eggLayer.AccumulatedFrametime += frameTime;
+            if (now < eggLayer.NextGrowth)
+                continue;
+
+            // Randomize next growth time for more organic egglaying.
+            eggLayer.NextGrowth = now + TimeSpan.FromSeconds(_random.NextFloat(eggLayer.EggLayCooldownMin, eggLayer.EggLayCooldownMax));
 
-            if (eggLayer.AccumulatedFrametime < eggLayer.CurrentEggLayCooldown)
+            if (_mobState.IsDead(uid))
```
Comment (metalgearsloth) [inline Content.Server/Animals/Systems/EggLayerSystem.cs:56]: This is going to eat the remainder after every tick, you should just be adding the timespan to the time on nextgrowth and not currenttime.

File: Content.Shared/Animal/UdderComponent.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Content.Shared.Animals;
+
+/// <summary>
+///     Gives the ability to produce a solution;
+///     produces endlessly if the owner does not have a HungerComponent.
+/// </summary>
+[RegisterComponent]
+[AutoGenerateComponentPause]
+public sealed partial class UdderComponent : Component
```
Comment (metalgearsloth) [inline Content.Shared/Animal/UdderComponent.cs:14]: Any reason it's not networked?

File: Content.Shared/Animal/UdderComponent.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+{
+    /// <summary>
+    ///     The reagent to produce.
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadOnly)]
+    public ProtoId<ReagentPrototype> ReagentId = default!;
+
```
Comment (metalgearsloth) [inline Content.Shared/Animal/UdderComponent.cs:21]: As above do not use default! for datafields.

File: Content.Shared/Animal/UdderComponent.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    /// </summary>
+    [DataField]
+    public TimeSpan GrowthDelay = TimeSpan.FromMinutes(1);
+
+    /// <summary>
+    ///     When to next try to produce.
+    /// </summary>
+    //[DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), Access(typeof(SharedUdderSystem))]
+    //public TimeSpan NextGrowth = TimeSpan.FromSeconds(0);
+    [AutoPausedField, Access(typeof(UdderSystem))]
```
Comment (metalgearsloth) [inline Content.Shared/Animal/UdderComponent.cs:56]: Please remove dummy code also no datafield means you'll break map saving.

File: Content.Shared/Animal/WoolyComponent.cs [hunk 14]
```diff
@@ -35,24 +34,26 @@ public sealed partial class WoolyComponent : Component
...
 
     /// <summary>
     ///     When to next try growing wool.
     /// </summary>
-    [DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), ViewVariables(VVAccess.ReadWrite)]
-    public TimeSpan NextGrowth = TimeSpan.FromSeconds(0);
+    //[DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), Access(typeof(SharedWoolySystem))]
+    //public TimeSpan NextGrowth = TimeSpan.FromSeconds(0);
+    [AutoPausedField, Access(typeof(WoolySystem))]
+    public TimeSpan NextGrowth = TimeSpan.Zero;
```
Comment (metalgearsloth) [inline Content.Shared/Animal/WoolyComponent.cs:58]: As above.

File: Content.Shared/Animal/WoolySystem.cs.cs [hunk 15]
```diff
@@ -1,29 +1,29 @@
...
-    [Dependency] private readonly SolutionContainerSystem _solutionContainer = default!;
+    [Dependency] private readonly SharedSolutionContainerSystem _solutionContainer = default!;
 
     public override void Initialize()
     {
         base.Initialize();
 
         SubscribeLocalEvent<WoolyComponent, BeforeFullyEatenEvent>(OnBeforeFullyEaten);
+        Logger.Info("Subscribed to SharedBeforeFullyEatenEvent for SharedWoolyComponent");
     }
```
Comment (metalgearsloth) [inline Content.Shared/Animal/WoolySystem.cs.cs:27]: This log should not exist nor should this need to be a message as subscriptions are entirely unordered.

File: Content.Shared/Animal/WoolySystem.cs.cs [hunk 16]
```diff
@@ -52,16 +58,14 @@ public override void Update(float frameTime)
...
-
             _solutionContainer.TryAddReagent(wooly.Solution.Value, wooly.ReagentId, wooly.Quantity, out _);
         }
     }
 
     private void OnBeforeFullyEaten(Entity<WoolyComponent> ent, ref BeforeFullyEatenEvent args)
     {
         // don't want moths to delete goats after eating them
+        Logger.Info($"OnBeforeFullyEaten called for entity {ent}");
         args.Cancel();
```
Comment (metalgearsloth) [inline Content.Shared/Animal/WoolySystem.cs.cs:69]: As above.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Animals/Systems/EggLayerSystem.cs (hunk 10 above)
Comment (Sirionaut) [reply Content.Server/Animals/Systems/EggLayerSystem.cs:56]: so I make it `eggLayer.NextGrowth += TimeSpan.FromSeconds(_random.NextFloat(eggLayer.EggLayCooldownMin, eggLayer.EggLayCooldownMax));` right?

Should I do the same in UdderSystem and WoolySystem?

File: Content.Server/Animals/Components/EggLayerComponent.cs (hunk 9 above)
Comment (Sirionaut) [reply Content.Server/Animals/Components/EggLayerComponent.cs:41]: Yeah, I realize it's against the guidelines.... it did start out as a tweak and while working on it, I realized there is a bug.

Should I remove those balance changes from here and make it a separate PR?

For reference, the cooldown used to be between 60s and 120s, now it would be 45s and 90s.
The more impactful balance change is the nutrient cost, which was 60 for one egg, now 25. Compared to milk which is 10 nutrient for 25u.

File: Content.Shared/Animal/UdderComponent.cs (hunk 11 above)
Comment (Sirionaut) [reply Content.Shared/Animal/UdderComponent.cs:14]: The reason is that they weren't before, and I only understand half of why to do certain things, when it comes to the whole networking/shared thing.

Looking at other components on shared, I now suspect almost all components on shared should be [NetworkedComponent] ? So I add it to all the components in the PR here?

Comment (Tayrtahn) [reply Content.Shared/Animal/UdderComponent.cs:14]: Yep. [NetworkedComponent] makes it so that adding or removing the component on the server does the same on the client, which we want here.

**Comment by Sirionaut** [comment]
Reverted the balance change, which will be its own PR after this one is merged.

File: Content.Shared/Animal/UdderComponent.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+using Content.Shared.FixedPoint;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Animals;
+
+/// <summary>
+///     Gives the ability to produce a solution;
+///     produces endlessly if the owner does not have a HungerComponent.
+/// </summary>
+[RegisterComponent, AutoGenerateComponentPause]
```
Comment (Tayrtahn) [inline Content.Shared/Animal/UdderComponent.cs:12]: NetworkedComponent

File: Content.Shared/Animal/WoolyComponent.cs [hunk 18]
```diff
@@ -1,17 +1,15 @@
...
+namespace Content.Shared.Animals;
 
 /// <summary>
-///     Lets an entity produce wool fibers. Uses hunger if present.
+///     Gives the ability to produce wool fibers;
+///     produces endlessly if the owner does not have a HungerComponent.
 /// </summary>
-
-[RegisterComponent, Access(typeof(WoolySystem))]
+[RegisterComponent, Access(typeof(WoolySystem)), AutoGenerateComponentPause]
```
Comment (Tayrtahn) [inline Content.Shared/Animal/WoolyComponent.cs:12]: NetworkedComponent

**Comment by slarticodefast** [comment]
Closing this in favor of ä32905 which fixed the merge conflict and addressed the new review there.
Thanks to you and Baa for fixing this, and sorry for the long wait.
